### PR TITLE
Build mod files a-posteriori (akin to nrnivmodl)

### LINKED ISSUE #38: nrnivmol_core for coreneuron (nrnivmodl equivalent)
All NEURON users are familiar with building special using `nrnivmodl`.
When we build neuron, `nrnivmodl` script is generate which takes care of:
* fining all mod files and translating via nocmodl
* create shared / static library
* link with necessary neuron library and create `special`

Similar workflow should be available with CoreNEURON. We should create `nrnivmodl_core` as part of CMake installation.

@fouriaux : when you install neuron, check nrnivmodl script under `bin`. We have to use [configure_file](https://cmake.org/cmake/help/v3.2/command/configure_file.html) functionality of CMake.

### PR DESCRIPTION
**Pull request by ferdonline**
This PR implements the required changes to build mod files later, just like neuron. 
With this change only one "base" coreneuron shall be built in a deployment, and greatly simplifies deploying multiple models to run with coreneuron.

Building groups of mechanisms is started by a `nrnivmodl-core` script. It will create a folder named `<ARCH>`, e.g. "x86_64" in the current dir with the contents:
- a library in the format `libcorenrnmech<_SUFFIX>-<VERSION>.[so/dynlib]` 
- a `special-core` runnable linked with the previous lib
- `modc_core` folder with the mods after conversion to c using mod2c
- `build_core` folder with temporary object files

Besides the mods c source, modc_core will contain as well sources of auxiliary files that depend on the mods: `_mod_func.cpp`, `_kinderiv.h` and `_dimplic.cpp`

Script options are available with `-h`: 
```
[leite@bbpv2 CoreNeuron]$ bin/nrnivmodl-core -h
nrnivmodl-core [options, ...] [mods_path]
Options:
  -n <name>       Optionally add mech name to library, as suffix
  -v <version>    Set libcorenrnmech version (default: 0.0)
  -o <output_dir> Set the output dir (default: <arch>-core)
  -i <incl_flags> Definitions passed to the compiler, typically '-I dir..'
  -l <link_flags> Definitions passed to the linker, typically '-Lx -lylib..'
  -c              (clean) Don't use any ADDITIONAL mod built-into coreneuron.
                  By default this only contains halfgap.mod
```

While most options are obvious, `-c` is something new. Coreneuron can be compiled with additional mods built-in (by default halfgap.mod, but configurable using Cmake's ADDITIONAL_MECHS var). When `-c` is specified it won't add these mods to `_mod_func` and therefore they are not initialized.

Fixes #38

### CODE REVIEW

**Comment by pramodk** [comment]
> Can one of the admins verify this patch?

@ferdonline  : I don't understand how you are not-recognised as BBP member :)

File: coreneuron/CMakeLists.txt [hunk 1]
```diff
@@ -203,7 +203,7 @@ mod2c_from_file(OPTMECH "${ADDITIONAL_MECHS}" "${ADDITIONAL_MECHPATH}")
 # For 'non-standard' mod files, need to generate registering function in mod_func.c
 
 set(MOD_FUNC_C    "${CMAKE_CURRENT_BINARY_DIR}/mod_func.cpp")
-set(MOD_FUNC_C_PL "${CMAKE_CURRENT_SOURCE_DIR}/mech/mod_func.c.pl")
+set(MOD_FUNC_C_PL "${CMAKE_CURRENT_SOURCE_DIR}/mech/mod_func.c.pl" CACHE STRING "Create mod registration c source")
```
Comment (ferdonline) [inline coreneuron/CMakeLists.txt]: I guess this can be reverted

**Comment by ferdonline** [comment]
@pramodk I fixed mac and some more bugs. It seems to be working now, at least in my setup.

File: extra/corenrnmodl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+
+make_params=()
+for param in MECH_NAME MECH_VERSION OUTPUT INCFLAGS LINKFLAGS MODS_PATH; do
+    var="params_${param}"
+    [ "${!var}" ] && make_params+=("$param=${!var}")
+done
+
+# Fail early if no mods
+ls "${params_MODS_PATH:-.}"/*.mod || { echo "No mods found"; exit 1; }
```
Comment (pramodk) [inline extra/corenrnmodl]: doesn't neuron fail if there is no mod file provided? I don't remember...

File: coreneuron/CMakeLists.txt (hunk 1 above)
Comment (ferdonline) [reply coreneuron/CMakeLists.txt]: I can fix this

File: apps/CMakeLists.txt [hunk 3]
```diff
@@ -30,14 +30,18 @@ include_directories(${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/coreneuron ${CMA
...
 
+# LIbrary required for special
+add_library(coreneuron_main STATIC "coreneuron.cpp")
+
 add_executable(coreneuron_exec "${CMAKE_CURRENT_SOURCE_DIR}/coreneuron.cpp")
 
 set_target_properties(coreneuron_exec PROPERTIES
-                            OUTPUT_NAME "coreneuron_exec"
-                            RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin")
+                      OUTPUT_NAME "coreneuron_exec")
```
[the remark this answers is not in the document]
Comment (ferdonline) [reply apps/CMakeLists.txt]: Go ahead with the renaming

File: extra/corenrnmodl (hunk 2 above)
Comment (ferdonline) [reply extra/corenrnmodl]: I will also work do this

**Comment by pramodk** [comment]
Please retest

**Comment by pramodk** [comment]
@ferdonline  : I was about to merge and while testing realise last small thing:

In `coreneuron/CMakeLists.txt`:

```
# to work around no acc ability to pass function pointers as arguments,
# some translated c files depend on a _kinderiv.h file that is constructed
# by kinderiv.py

set(KINDERIV_PY "${CMAKE_CURRENT_SOURCE_DIR}/kinderiv.py")
set(KINDERIV_H "${CMAKE_CURRENT_BINARY_DIR}/_kinderiv.h")

add_custom_command(OUTPUT "${KINDERIV_H}"
    DEPENDS ${MOD2C_OPTMECH_OUTPUTS} ${MOD2C_STDMECH_OUTPUTS} "${KINDERIV_PY}"
    COMMAND python "${KINDERIV_PY}"
    WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}")
```

So we basically need to re-run above python script after translation of mod to c. 

We can reproduce the error with:

```
git clone ssh://bbpcode.epfl.ch/user/kumbhar/reduced_dentate

install/bin/nrnivmodl-core reduced_dentate/mechanisms

/usr/local/bin/mpicxx -fPIC -DNRNMPI=1 -DLAYOUT=0 -DDISABLE_HOC_EXP -DENABLE_SPLAYTREE_QUEUING -DCORENEURON_BUILD -I/Users/kumbhar/workarena/repos/bbp/coreneuron/b/install/include -I/Users/kumbhar/workarena/repos/bbp/coreneuron/b/install/include/coreneuron/utils/randoms  -c modc_core/Aradi_Ca.cpp -o Aradi_Ca.mod.o
modc_core/Aradi_Ca.cpp:494:68: error: use of undeclared identifier '_derivimplicit_state_Aradi_Ca'
modc_core/Aradi_CadepK.cpp:432:68: _reset = nrn_newton_thread((NewtonSpace*)_newtonspace1, 6,_slist2, _derivimplicit_state_Aradi_Ca, _dlist2,  _threadargs_);error
: use of undeclared identifier '_derivimplicit_state_CadepK'
_reset = nrn_newton_thread((NewtonSpace*)_newtonspace1, 4,_slist2, _derivimplicit_state_CadepK, _dlist2,  _threadargs_);
```
I release that we are installing `install/include/_kinderiv.h` (this shouldn't go into include dir)

lets sync tomorrow.

**Comment by ferdonline** [comment]
@pramodk anything missing here? Notice I also included the patch for the reporting lib size.

**Comment by ferdonline** [comment]
Version which works well also with plasticity. Time to run final tests

**Comment by pramodk** [comment]
Please retest

**Comment by ferdonline** [comment]
Some documentation on the makefile: [coreneuron_nrnivmodl.pdf](https://github.com/BlueBrain/CoreNeuron/files/2840483/coreneuron_nrnivmodl.pdf)

**Comment by ferdonline** [comment]
@pramodk do you understand why the tests are timing out?

**Comment by ferdonline** [comment]
retest this please

**Comment by ferdonline** [comment]
retest this please

**Comment by ferdonline** [comment]
retest this please

**Comment by ferdonline** [comment]
retest this please

**Comment by ferdonline** [comment]
retest this please

**Comment by pramodk** [comment]
thank you @ferdonline !!!!

**Pull request merged**